### PR TITLE
Cosmos integration tests

### PR DESCRIPTION
**Pull request by margaret**
## High Level Description
- Add integration test (add and remove kafka) for packaging using new `dcos_session_api` cosmos support. 
- Moved the pkgpanda integration test as well as the cosmos integration test to a separate test_packaging module
- Skip test if there are not enough resources on the cluster
- I left `log.debug`s in, lmk if I should remove those.

## Related Issues

  - [DCOS_OSS-1394](https://jira.mesosphere.com/browse/DCOS_OSS-1394) Add/remove package support in dcos-test-utils
  - This one will support [DCOS_OSS-1243](https://jira.mesosphere.com/browse/DCOS_OSS-1243) Add Tweeter Demo to Integration Test

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: This PR is for the test, but the code it is testing lives in [dcos-test-utils](http://github.com/mesosphere/dcos-test-utils)
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

File: packages/dcos-integration-test/extra/test_applications.py [hunk 1]
```diff
@@ -222,3 +222,23 @@ def assert_packages_match_active_buildinfo(package_ids):
 
         assert set(active_package_ids) <= set(package_ids)
         assert_packages_match_active_buildinfo(active_package_ids)
+
+
+def test_packaging_api(dcos_api_session):
+    """Test the Cosmos API (/package) wrapper
+    """
+    install_response = dcos_api_session.cosmos.install_package('kafka', '1.1.9-0.10.0.0')
```
Comment (mellenburg) [inline packages/dcos-integration-test/extra/test_applications.py]: What are the minimum system requirements for launching kafka? We probably want to decorate this test with something that will either skip or fail if the cluster can't possibly support this test.

FYI current PR integration test cluster config is 2 private and 1 public usually. M3.xlarge in provider: aws and m4.xlarge in provider: onprem

Comment (margaret) [inline packages/dcos-integration-test/extra/test_applications.py]: Ah good point. It worked when I manually ran the current version of this PR, so it seems like 2 is fine but I'll check.

Would it be a problem increasing it to the required 5 agents for the tweeter integration test? Seems kind of wasteful if there's only the one integration test that uses it

Comment (mellenburg) [inline packages/dcos-integration-test/extra/test_applications.py]: What we will do at that time is spin up just one cluster with the 5 agents, run tweeter (and nothing else) against that, and then tear it down. All the other tests will ignore tweeter and still use the more minimal cluster.

Comment (margaret) [inline packages/dcos-integration-test/extra/test_applications.py]: The most recent version of [kakfa](https://github.com/mesosphere/universe/blob/version-3.x/repo/packages/K/kafka/39/config.json#L97) says it needs 3 brokers, and each broker needs 1.0 cpus with 2304 mem, 2048 heap, and 5000 disk.

**Comment by margaret** [comment]
AWS failure is `test_rexray.test_move_external_volume_to_new_agent` and code-quality is related to the unpinned flake8 issue that would be fixed by https://github.com/dcos/dcos/pull/1776

**Comment by margaret** [comment]
@mesosphere-mergebot bump-ee

File: packages/dcos-integration-test/extra/test_packaging.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    },
+    'executor': {
+        'disk': 0,
+        'mem': 256,
+        'cpus': 0.5
+    }
+}
+
+
+def get_cluster_resources(dcos_api_session):
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_packaging.py]: private? `_get_cluster_resources` ? 
Actually, given the single usage. My tendency is to inline it at the place it is used `dcos_api_session.get('mesos/state-summary').json()`

File: packages/dcos-integration-test/extra/test_packaging.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    """Return the mesos state summary
+    """
+    r = dcos_api_session.get('mesos/state-summary')
+    return r.json()
+
+
+def enough_resources_for_package(state_summary, package_requirements):
+    """SANITY CHECK that there are enough unreserved_resources in a state to run a package.
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_packaging.py]: why SANITY CHECK all caps? Title Case could be better IMO.

Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_packaging.py]: also `_private` ?

File: packages/dcos-integration-test/extra/test_packaging.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        if usable_nodes == package_requirements['number_of_nodes']:
+            return True
+    log.info('Only {available} usable agents. This package needs {required} with these resources {resources}.'.format(
+        available=usable_nodes,
+        required=package_requirements['number_of_nodes'],
+        resources=package_requirements['node']))
+    return False
+
+
+def agent_has_resources(agent, node_requirements):
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_packaging.py]: If we go by definition follows use. This function should be _private and should be defined above _enough_resources_for_package.

File: packages/dcos-integration-test/extra/test_packaging.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    unreserved = agent['unreserved_resources']
+    resources = ['mem', 'disk', 'cpus']
+    enough_resources = True
+    for resource in resources:
+        log.debug('{resource}: unreserved {unreserved}, required {required}'.format(
+            resource=resource,
+            unreserved=unreserved[resource],
+            required=node_requirements[resource]))
+        if unreserved[resource] < node_requirements[resource]:
+            enough_resources = False
```
Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_packaging.py]: Could this return early? If e.g. if men is not sufficient, we could just return `False` instead of checking disk and cpus.

File: packages/dcos-integration-test/extra/test_packaging.py (hunk 3 above)
Comment (margaret) [inline packages/dcos-integration-test/extra/test_packaging.py]: I was planning to use both this and get_cluster_resources in the tweeter integration test, but that's going to be a separate PR

Comment (orsenthil) [inline packages/dcos-integration-test/extra/test_packaging.py]: I see. I think, it could still be made `_private` to denote that these are helper functions for certain tests.

File: packages/dcos-integration-test/extra/test_packaging.py (hunk 5 above)
Comment (margaret) [inline packages/dcos-integration-test/extra/test_packaging.py]: Yeah, that would be more efficient. I can change it. I wrote it like this to get the log.debug statements for all of them since there are only three items to check anyway.

**Comment by margaret** [comment]
@mesosphere-mergebot bump-ee

**Comment by margaret** [comment]
@mesosphere-mergebot bump-ee

**Comment by margaret** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by margaret** [comment]
Looks like smoke-test failure is https://jira.mesosphere.com/browse/DCOS_OSS-1481

**Comment by orsenthil** [comment]
@margaret  - looks like error caught by the test.

```
>               log.debug('Agent does not have has enough {}'.format())
E               IndexError: tuple index out of range

test_packaging.py:88: IndexError
```

**Comment by margaret** [comment]
Btw I have manually tested this one by upping the number of agents in the teamcity parameters and re-running, in which case it won't skip these new tests. Eventually going to set up a separate job for this and demo integration tests and anything else that requires installing apps onto more than 2 private nodes.

**Comment by margaret** [comment]
@mesosphere-mergebot bump-ee

**Comment by margaret** [comment]
@orsenthil are you or @karlkfi working on the dcos-docker 'no space left' error?

aws-deploy failure is the rexray test
dcos-docker failures look like they're both caused by this bug [smoke-test/dcos-docker: 'no space left on device'](https://jira.mesosphere.com/browse/DCOS_OSS-1481)

**Comment by margaret** [comment]
The aws-deploy failure is the dependably flaky rexray test

**Comment by margaret** [comment]
@mesosphere-mergebot bump-ee

File: packages/dcos-integration-test/extra/test_packaging.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+
+def test_packaging_api(dcos_api_session):
+    """Test the Cosmos API (/package) wrapper
+    """
+    _skipif_insufficient_resources(dcos_api_session, KAFKA_PACKAGE_REQUIREMENTS)
+    install_response = dcos_api_session.cosmos.install_package('kafka', '1.1.9-0.10.0.0')
+    data = install_response.json()
+
+    dcos_api_session.marathon.poll_marathon_for_app_deployment(data['appId'], 1,
```
Comment (mellenburg) [inline packages/dcos-integration-test/extra/test_packaging.py]: This is inconsequential, but instead of polling for this specific application you can just broadly wait for all deployments to end with this: https://github.com/dcos/dcos-test-utils/blob/master/dcos_test_utils/marathon.py#L298 and then have a much simpler function signature

**Comment by margaret** [comment]
@mesosphere-mergebot bump-ee
